### PR TITLE
8297072: Provide gradle option to test a previously built SDK

### PR DESCRIPTION
**Pull request by arapte**
#### Requirement:
We want to be able to test a JavaFX SDK built somewhere else other than on current machine or current repo on same machine. Which means we should be able to run the unit and system tests with a different JavaFX SDK.

##### Run tests using a specified JavaFX SDK:
- Introduce a property `TEST_JAVAFX_SDK_PATH` that points to the JavaFX SDK which was built elsewhere.
- Modify the paths generated by build.gradle to point to the path sepcified by `TEST_JAVAFX_SDK_PATH`
- When a `TEST_JAVAFX_SDK_PATH` is specified `-PTEST_ONLY=true` must also be specified.

##### Generate the JavaFX SDK bundle
- Generate the minimal JavaFX SDK bundle that is required to run the tests: We require content of the _build/sdk_ and _build/shims_ directories to achieve the same.

##### With a single machine:
1. Create two local copies JavaFX repo, [repo1, repo2]
2. Build sdk and shims in repo1. `gradle sdk shims`
3. In repo2, run different tests as,
> gradle -PTEST_ONLY=true -PTEST_JAVAFX_SDK_PATH=**repo1/build**  :base:test <any tests>
4. In repo1 run `gradle all`
5. `javafx-sdk-shims.zip` file would be created under repo1/build. unzip it any `unzip-path` on machine
6. In repo2, run the tests by providing the `unzip-path` as TEST_JAVAFX_SDK_PATH
> gradle -PTEST_ONLY=true -PTEST_JAVAFX_SDK_PATH=**unzip-path**  :base:test <any tests>

##### With 2 machines.
1. Run `gradle all` on a machine1
2. Move build/javafx-sdk-shims.zip to machine2 and unzip to any `unzip-path`
3. Run the tests by providing the `unzip-path` as TEST_JAVAFX_SDK_PATH
> gradle -PTEST_ONLY=true -PTEST_JAVAFX_SDK_PATH=**unzip-path**  :base:test <any tests>

#### Additional changes:
Additionally we observed that three unit tests failed because they access a specific path. A path that would no exists when JavaFX SDK is not built.
1. Controls test: test.javafx.scene.control.ControlTest.testRT18097
It required path: modules/javafx.controls/build/classes/java/main/javafx.controls/javafx
It is fixed by change on line 2811

2. Graphics test: test.javafx.css.CssMetaDataTest.testRT18097
It required path: modules/javafx.graphics/build/classes/java/main/javafx.graphics/javafx
It is fixed by change on line 2722

3. Base test: test.com.sun.javafx.runtime.VersionInfoTest
It required path: build/module-lib/javafx.properties
VersionInfoTest is specific to the local build of JavaFX. So, It can be safely excluded when TEST_JAVAFX_SDK_PATH is specified.
It is fixed by change on line 2304 to 2310

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8297072](https://bugs.openjdk.org/browse/JDK-8297072): Provide gradle option to test a previously built SDK (**Enhancement** - P3)

### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Lukasz Kostyra](https://openjdk.org/census#lkostyra) (@lukostyra - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1577/head:pull/1577` \
`$ git checkout pull/1577`

Update a local copy of the PR: \
`$ git checkout pull/1577` \
`$ git pull https://git.openjdk.org/jfx.git pull/1577/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1577`

View PR using the GUI difftool: \
`$ git pr show -t 1577`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1577.diff">https://git.openjdk.org/jfx/pull/1577.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1577#issuecomment-2368704878)

### CODE REVIEW

**Review by kevinrushforth — COMMENTED** [review]
I tried the first recipe, with repo1 pointing to a full build of the current jfx master, and repo2 pointing to a clean repo with the patch from this PR. I deliberately did not build the sdk in repo2.

I was able to compile and run the tests from all projects except `javafx.swt`. Those tests are disabled (not run), but compilation of the tests is enabled, and it fails without having first compiled the "sdk" in repo2. Here is the error:

```
$ cd repo2
$ gradle --info --continue -PTEST_ONLY=true -PTEST_JAVAFX_SDK_PATH=repo1/build test
...
> Task :swt:compileTestJava FAILED
Caching disabled for task ':swt:compileTestJava' because:
  Build cache is disabled
  'Forking compiler via ForkOptions.executable' satisfied
Task ':swt:compileTestJava' is not up-to-date because:
  Task has failed previously.
The input changes require a full rebuild for incremental task ':swt:compileTestJava'.
Compilation mode: command line compilation
Compiling with toolchain '/jdk-22.0.2.jdk/Contents/Home'.
Compiling with Java command line compiler '/jdk-22.0.2.jdk/Contents/Home/bin/javac'.
Starting process 'command '/jdk-22.0.2.jdk/Contents/Home/bin/javac''. Working directory: jfx/modules/javafx.swt Command: /jdk-22.0.2.jdk/Contents/Home/bin/javac @jfx/modules/javafx.swt/build/tmp/compileTestJava/java-compiler-args.txt
Successfully started process 'command '/jdk-22.0.2.jdk/Contents/Home/bin/javac''
jfx/modules/javafx.swt/src/test/java/test/javafx/embed/swt/FXCanvasScaledTest.java:43: error: package javafx.embed.swt does not exist
import javafx.embed.swt.FXCanvas;
                       ^
jfx/modules/javafx.swt/src/test/java/test/javafx/embed/swt/FXCanvasScaledTest.java:123: error: cannot find symbol
    private static void initFX(FXCanvas canvas) {
                               ^
  symbol:   class FXCanvas
  location: class FXCanvasScaledTest
jfx/modules/javafx.swt/src/test/java/test/javafx/embed/swt/SWTCursorsTest.java:28: error: package javafx.embed.swt does not exist
import javafx.embed.swt.FXCanvas;
                       ^
jfx/modules/javafx.swt/src/test/java/test/javafx/embed/swt/FXCanvasTest.java:28: error: package javafx.embed.swt does not exist
import javafx.embed.swt.FXCanvas;
                       ^
jfx/modules/javafx.swt/src/test/java/test/javafx/embed/swt/FXCanvasScaledTest.java:71: error: cannot find symbol
        final FXCanvas canvas = new FXCanvas(shell, SWT.NONE);
              ^
  symbol:   class FXCanvas
  location: class FXCanvasScaledTest
jfx/modules/javafx.swt/src/test/java/test/javafx/embed/swt/FXCanvasScaledTest.java:71: error: cannot find symbol
        final FXCanvas canvas = new FXCanvas(shell, SWT.NONE);
                                    ^
  symbol:   class FXCanvas
  location: class FXCanvasScaledTest
jfx/modules/javafx.swt/src/test/java/test/javafx/embed/swt/SWTCursorsTest.java:49: error: cannot find symbol
        final FXCanvas canvas = new FXCanvas(shell, SWT.NONE);
              ^
  symbol:   class FXCanvas
  location: class SWTCursorsTest
jfx/modules/javafx.swt/src/test/java/test/javafx/embed/swt/SWTCursorsTest.java:49: error: cannot find symbol
        final FXCanvas canvas = new FXCanvas(shell, SWT.NONE);
                                    ^
  symbol:   class FXCanvas
  location: class SWTCursorsTest
jfx/modules/javafx.swt/src/test/java/test/javafx/embed/swt/FXCanvasTest.java:47: error: cannot find symbol
        final FXCanvas canvas = new FXCanvas(shell, SWT.NONE);
              ^
  symbol:   class FXCanvas
  location: class FXCanvasTest
jfx/modules/javafx.swt/src/test/java/test/javafx/embed/swt/FXCanvasTest.java:47: error: cannot find symbol
        final FXCanvas canvas = new FXCanvas(shell, SWT.NONE);
                                    ^
  symbol:   class FXCanvas
  location: class FXCanvasTest
jfx/modules/javafx.swt/src/test/java/test/javafx/embed/swt/FXCanvasTest.java:55: error: cannot find symbol
        assertSame(canvas, FXCanvas.getFXCanvas(canvas.getScene()));
                           ^
  symbol:   variable FXCanvas
  location: class FXCanvasTest
11 errors
Resolve mutations for :swt:processTestResources (Thread[#65,included builds,5,main]) started.
:swt:processTestResources (Thread[#65,included builds,5,main]) started.

> Task :swt:processTestResources UP-TO-DATE
Caching disabled for task ':swt:processTestResources' because:
  Build cache is disabled
  Not worth caching
Skipping task ':swt:processTestResources' as it is up-to-date.

FAILURE: Build failed with an exception.
```

Unlike the other modules, the `javafx.swt` module is not loaded from the module path, so the absence of the sdk build in the repo you are testing causes the compilations of the javafx.swt tests to fail. Since they aren't being run, perhaps the best solution is to disable the compileTestJava task in the `:swt` project if `IS_TEST_JAVAFX_SDK` is true.

I left a minor naming suggestion for the name of the new gradle property.

File: build.gradle [hunk 1]
```diff
@@ -729,6 +729,24 @@ def buildSdkForTest = Boolean.parseBoolean(BUILD_SDK_FOR_TEST)
...
+    }
+    def sdkShimsPath = file("${TEST_JAVAFX_SDK_PATH}")
+    def sdkPath = file("${TEST_JAVAFX_SDK_PATH}/sdk")
+    def shimsPath = file("${TEST_JAVAFX_SDK_PATH}/shims")
+    if (!sdkShimsPath.exists() || !sdkShimsPath.isDirectory() ||
+             !sdkPath.exists() || !sdkPath.isDirectory() ||
+           !shimsPath.exists() || !shimsPath.isDirectory()) {
+        fail("The provided TEST_JAVAFX_SDK_PATH=${TEST_JAVAFX_SDK_PATH} is invalid")
+    }
+    ext.IS_TEST_JAVAFX_SDK = true
```
[the remark this answers is not in the document]
Comment (arapte) [reply build.gradle]: Changed name as `IS_TEST_SDK`

File: build.gradle [hunk 2]
```diff
@@ -729,6 +729,24 @@ def buildSdkForTest = Boolean.parseBoolean(BUILD_SDK_FOR_TEST)
 defineProperty("TEST_ONLY", "false")
 ext.IS_TEST_ONLY = Boolean.parseBoolean(TEST_ONLY) || !buildSdkForTest
 
+ext.IS_TEST_JAVAFX_SDK = false
+if (hasProperty("TEST_JAVAFX_SDK_PATH")) {
```
[the remark this answers is not in the document]
Comment (arapte) [reply build.gradle]: Changed name as `TEST_SDK_PATH`

**Comment by arapte** [comment]
>  disable the compileTestJava task in the :swt project if IS_TEST_JAVAFX_SDK is true.
disabled the task accordingly.

Thanks for the review, updated the PR as per comments. 
The names TEST_SDK_PATH and IS_TEST_SDK seem better.

**Review by kevinrushforth — APPROVED** [review]
Looks good. I ran the following in a clean repo1 and all tests passes:

```
gradle  -PTEST_ONLY=true -PFULL_TEST=true -PTEST_SDK_PATH=repo2/build test
```

**Comment by arapte** [comment]
@kevinrushforth As you guided offline.
Updated the PR to provide an option to generate the run.args and compile.args in cases when needed. Kindly re-review.

**Comment by kevinrushforth** [comment]
@lukostyra Can you be the second reviewer?

**Review by lukostyra — APPROVED** [review]
LGTM, tested single-machine on Windows with sdk-shims zip exported out. Good work!

**Pull request closed without merging**